paymentintent.go

```diff
@@ -0,0 +1,174 @@
+}
+
+// PaymentIntentParams is the set of parameters that can be used when handling a payment intent.
+type PaymentIntentParams struct {
+	Params               `form:"*"`
+	AllowedSourceTypes   []*string                        `form:"allowed_source_types"`
+	Amount               *int64                           `form:"amount"`
+	ApplicationFee       *int64                           `form:"application_fee"`
+	AttemptConfirmation  *string                          `form:"attempt_confirmation"`
+	CaptureMethod        *string                          `form:"capture_method"`
+	Currency             *string                          `form:"currency"`
+	Customer             *string                          `form:"customer"`
+	Description          *string                          `form:"description"`
+	OnBehalfOf           *string                          `form:"on_behalf_of"`
+	ReceiptEmail         *string                          `form:"receipt_email"`
+	ReturnURL            *string                          `form:"return_url"`
+	SaveSourceToCustomer *bool                            `form:"save_source_to_customer"`
+	Shipping             *ShippingDetailsParams           `form:"shipping"`
+	Source               *string                          `form:"source"`
+	StatementDescriptor  *string                          `form:"statement_descriptor"`
+	TransferData         *PaymentIntentTransferDataParams `form:"transfer_data"`
+	TransferGroup        *string                          `form:"transfer_group"`
+}
+
+// PaymentIntentListParams is the set of parameters that can be used when listing payment intents.
+// For more details see https://stripe.com/docs/api#list_payouts.
+type PaymentIntentListParams struct {
+	ListParams `form:"*"`
+}
+
+type PaymentIntentSourceActionAuthorizeWithUrl struct {
+	URL string `json:"url"`
+}
+
+// PaymentIntentSourceActionValue describes the `value` property in `next_source_action`
+// The `type` in the parent should indicate which object is fleshed out.
+type PaymentIntentSourceActionValue struct {
+	AuthorizeWithUrl *PaymentIntentSourceActionAuthorizeWithUrl `json:"-"`
+}
+
+type PaymentIntentSourceAction struct {
+	Type  PaymentIntentNextActionType     `json:"type"`
+	Value *PaymentIntentSourceActionValue `json:"value"`
+}
+
+// UnmarshalJSON handles deserialization of a PaymentIntentSourceAction.
+// This custom unmarshaling is needed because the specific
+// type of for `value` it refers to is specified in the `type` property
+func (s *PaymentIntentSourceAction) UnmarshalJSON(data []byte) error {
+	fmt.Printf("hello")
+
+	type paymentIntentSourceAction PaymentIntentSourceAction
+	var v paymentIntentSourceAction
+	if err := json.Unmarshal(data, &v); err != nil {
+		return err
+	}
+
+	var err error
+	*s = PaymentIntentSourceAction(v)
+
+	fmt.Printf("%v", s.Type)
+
+	switch s.Type {
+	case PaymentIntentNextActionAuthorizeWithURL:
+		fmt.Printf("%v", data)
+		err = json.Unmarshal(data, &s.Value.AuthorizeWithUrl)
```

Comment:
I know this part is incorrect because at that point `data` still contains the top-level info but I haven't yet figured out out I can fix it.